custom_components/gtfs_rt/StaticTimetable.py

```diff
@@ -0,0 +1,176 @@
+
+
+class GTFSCache:
+    """
+    Class for saving daily static GTFS data to a cache dictionary.
+    """
+
+    def __init__(self) -> None:
+        self.data = dict()
+
+    def save(self, date: datetime.date, data: pd.DataFrame) -> None:
+        """
+        Saves data to a dictionary for later access using "date" as a key.
+
+        :param date: The date to use as a dictionary key.
+        :type date: datetime.date
+        :param data: Data to be saved in the cache.
+        :type data: pd.DataFrame
+        """
+        _LOGGER.debug(f"Saving static GTFS data from {date} to cache...")
+        self.data = {date: data}
+
+    def load(self, date: datetime.date) -> pd.DataFrame | None:
+        """
+        Returns previously-cached data from the cache dictionary using "date"
+        as a key or else returns an empty dictionary if the date is not
+        present in the dictionary keys.
+
+        :param date: The date to use as a dictionary key.
+        :type date: datetime.date
+        :return: Previously-saved data from the cache dictionary or None.
+        :rtype: pd.DataFrame | None
+        """
+        try:
+            _LOGGER.debug(f"Loading cached static GTFS data from {date}...")
+            return self.data[date]
+        except KeyError:
+            _LOGGER.debug(f"No static GTFS data from {date} found in cache...")
+            return None
+
+
+class StaticMasterGTFSInfo:
+    """
+    Class to control access to the static GTFS information available from the
+    public transport service provider.
+    """
+
+    def __init__(self, url: str, CachedData: GTFSCache) -> None:
+        """
+        Creates the class object by first checking if there are cached data
+        available and returning those data. If there is no data for today's
+        date found in the cache, this class triggers downloading and creating
+        the required dictionary.
+
+        :param url: URL from which to download GTFS data.
+        :type url: str
+        :param CachedData: Previously-saved cache of data.
+        :type CachedData: GTFSCache
+        """
+        date_today = datetime.date.today()
+        cache = CachedData.load(date_today)
+        if cache is None:
+            _LOGGER.debug("Getting new static GTFS data...")
+            dataframes = get_dataframes(url)
+            cache = process_dataframes(dataframes)
+            CachedData.save(date_today, cache)
+        self.departure_info = cache
+
+
+def process_dataframes(dataframes: dict[str, pd.DataFrame]) -> pd.DataFrame:
+    """
+    Merges the provided Pandas dataframes as needed and return a single Stop
+    Schedule dataframe.
+
+    :param dataframes: Dictionary mapping filenames to Pandas dataframes.
+    :type dataframes: dict[str, pd.DataFrame]
+    :return: Merged dataframe of stop schedule information.
+    :rtype: pd.DataFrame
+    """
+
+    routes = dataframes["routes"]
+    trips = dataframes["trips"]
+    stop_times = dataframes["stop_times"]
+
```

Comment:
This constructs a timetable for the current day only and does not pay attention to whether the service runs on a particular day/date. 
- TODO: generate additional times. Maybe we can perform the caching option more frequently than daily and wrap the stop events that are in the past around to occur in the future. This would require adding a start_date parameter. 
- TODO: read calendar data in order to check if a service is running in the time period selected

Reply:
How to handle calendar_dates? 
Options:
- Create an Exceptions column which contains a dictionary for each route mapping date to exception type. Possibly slow to access but clean.
- Reshape calendar_dates to have a column for every date listed, creating multiple exception dates to check. Messy but fast to access (though would require a check to see if column is present).

Reply:
I went with option 2 and will filter that in the PublicTransportData class on each sensor update.